### DISCUSSION TITLE
Ambiguous resolution of dynamic routes catch all with another same level dynamic route on export

### DISCUSSION THREAD

**Discussion by aendel** [vercel/next.js#37845] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T8101
    Binaries:
      Node: 16.13.1
      npm: 8.3.0
      Yarn: 1.22.18
      pnpm: 7.1.7
    Relevant packages:
      next: 12.1.7-canary.38
      react: 17.0.2
      react-dom: 17.0.2

### What browser are you using? (if relevant)

Version 102.0.5005.115 (arm64)

### How are you deploying your application? (if relevant)

next export

### Describe the Bug

<img width="235" alt="Schermata 2022-06-14 alle 15 52 13" src="https://user-images.githubusercontent.com/12732872/173594089-880e4220-333e-41c6-ad25-6f202702ecda.png">

The previous image show my pages folder structure in which I have the following dynamics routes

**(be carefull that i'm aware of the [dynamics routes' caveats](https://nextjs.org/docs/routing/dynamic-routes#caveats))**

- [...catch-all]
- [slug]
- - [deep]

I'm always sure that where is no overlapping with catch-all, slug or deep paths.

```sh
Page                                       Size     First Load JS
┌ ○ /                                      2.7 kB         77.7 kB
├   └ css/d94d137a14864d9c.css             684 B
├   /_app                                  0 B            72.8 kB
├ ● /[...catch-all]                        438 B          75.4 kB
├   ├ /One/one
├   ├ /two-two
├   └ /Three
├ ● /[slug]                                412 B          75.4 kB
├   ├ /four
├   ├ /five
├   └ /six
├ ● /[slug]/[deep]                         450 B          75.4 kB
├   ├ /four/four-four
├   ├ /five/five-five
├   └ /six/six-six
├ ○ /404                                   2.97 kB        75.7 kB
└ λ /api/hello                             0 B            72.8 kB
+ First Load JS shared by all              72.8 kB
  ├ chunks/framework-e70c6273bfe3f237.js   42 kB
  ├ chunks/main-2a48abc85e3b6a19.js        28.6 kB
  ├ chunks/pages/_app-50ea61a192ce43b2.js  555 B
  ├ chunks/webpack-74ba1aba2cfe5288.js     1.57 kB
  └ css/27d177a30947857b.css
```

All the pages I expect have been exported properly.

But what happen is that when i try to go (through the next/link component or through the next/router functions) to the page /One/one this is going to render the [slug][deep] page component instead of the [...catch-all] page component.

It's happing the same with "two-two" with is going to be rendered with the [slug] component page.

**BUT. The catch is that when I try to reach the /One/one page through the url on the browser or through an <a> tag with the href the page is going to render the "expected" [...catch-all] page component.**

https://user-images.githubusercontent.com/12732872/173597525-57dda918-ddaa-4dbe-9dc1-29655587f79b.mov

### Expected Behavior

I would expect the next/link or next/router components/functions to behave as like as the anchor tag with the href behave in order to prevent mismatched client side page render.

### To Reproduce

I have created a repository with the minimal example in order to reproduce the bug, its's the same code which is shown in the screenshot and the video.

https://github.com/aendel/nextjs-catch-all-proof

In order to reproduce it:
1. Clone the repo: `git clone https://github.com/aendel/nextjs-catch-all-proof.git`
2. Install the dependencies: `npm install`
3. Build - Export - Serve static output: `npm run export:serve`
4. Follow the video at the end of the section "Describe the bug"

**Comment by ijjk** [comment]
Hi, on the client we don't have a list of all of the paths that are returned from `getStaticPaths` as this could potentially be thousands of paths which we don't want to require the client to load so the automatic resolving we do for `href` is best effort. 

We are investigating updated resolving with `middleware` and could potentially have a more verbose manifest for the client to consume when using `next export` as the server won't be available for that case. 

As a workaround for this currently you can use manual `href/as` values:

```js
<Link href='/[slug]/[deep]' as='/one/two'>
 <a>to /one/two</a>
</Link>
```

**Comment by aendel** [comment]
I would love to use the workaround that are you suggesting, but, in the actual application I'm developing, all my href come from a cms, which you can only write an href without any knowledge of which page is going to be render.

I was wondering if instead of returning alla the paths from the getStaticPaths, in the case of the export, we could use the prefetched paths in the page.

Another thought I had: We could specify in the page data json when __N_SSG":true" (or always) along with the pageProps also the slug of the page itself at build time in order to match it from the ssgManifest.
